flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp

```diff
@@ -617,6 +617,20 @@ DebugTypeGenerator::convertType(mlir::Type Ty, mlir::LLVM::DIFileAttr fileAttr,
     return convertRecordType(recTy, fileAttr, scope, declOp);
   } else if (auto tupleTy = mlir::dyn_cast_if_present<mlir::TupleType>(Ty)) {
     return convertTupleType(tupleTy, fileAttr, scope, declOp);
+  } else if (auto classTy = mlir::dyn_cast_if_present<fir::ClassType>(Ty)) {
```

Comment:
ClassType and BoxType are implemented in the same way in the runtime, the difference is that the dynamic type inside a ClassType may mismatch its static type.

You can actually merge the handling of ClassType and BoxType by replacing `auto boxTy = mlir::dyn_cast_or_null<fir::BoxType>(Ty)` with `auto boxTy = mlir::dyn_cast_or_null<fir::BaseBoxType>(Ty)`.

You would then get the static type description of the entity in the DWARF (which seems to be what is being done here).

Reply:
Thanks for your comments. I updated the patch to handle it with the BoxType. I also made another small change to handle cases like `CLASS(*)`.